go/pkg/services/log_collector/server.go

```diff
@@ -16,36 +16,507 @@ package log_collector
+
+	// get log file path
+	filePath, err := lcs.getLogFilePath(request.RunId)
 	if err != nil {
-		return nil, errors.Wrap(err, "Failed to create abstract server")
+		err := errors.Wrapf(err, "Failed to get log file path for run id %s", request.RunId)
+		return &log_collector.GetLogResponse{
+			Success: false,
+			Error:   err.Error(),
+		}, err
+	}
+
+	// open log file
+	file, err := os.Open(filePath)
+	if err != nil {
+		err := errors.Wrapf(err, "Failed to open log file for run id %s", request.RunId)
+		return &log_collector.GetLogResponse{
+			Success: false,
+			Error:   err.Error(),
+		}, err
+	}
+	defer file.Close() // nolint: errcheck
+
+	// get file size
+	fileInfo, err := file.Stat()
+	if err != nil {
+		err := errors.Wrapf(err, "Failed to get file info for run id %s", request.RunId)
+		return &log_collector.GetLogResponse{
+			Success: false,
+			Error:   err.Error(),
+		}, err
+	}
+
+	offset, size := lcs.validateOffsetAndSize(request.Offset, request.Size, uint64(fileInfo.Size()))
+	if size == 0 {
+		lcs.Logger.DebugWithCtx(ctx, "No logs to return", "run_id", request.RunId)
+		return &log_collector.GetLogResponse{
+			Success: true,
+			Log:     []byte{},
+		}, nil
 	}
-	return &logCollectorServer{
-		abstractServer,
+
+	// read size bytes from offset
+	buffer := make([]byte, size)
+	_, err = file.ReadAt(buffer, int64(offset))
+	if err != nil {
+
+		// if error is EOF, return empty bytes
+		if err == io.EOF {
+			return &log_collector.GetLogResponse{
+				Success: true,
+				Log:     buffer,
+			}, nil
+		}
+
+		// else return error
+		err := errors.Wrapf(err, "Failed to read log file for run id %s", request.RunId)
+		return &log_collector.GetLogResponse{
+			Success: false,
+			Error:   err.Error(),
+		}, err
+	}
+
+	return &log_collector.GetLogResponse{
+		Success: true,
+		Log:     buffer,
 	}, nil
 }
+
+// streamPodLogs streams logs from a pod and writes them into a file
+func (lcs *LogCollectorServer) streamPodLogs(ctx context.Context, runId, podName string) {
+
+	// create a log file to the pod
+	filePath := lcs.resolvePodLogFilePath(runId, podName)
+	if err := common.EnsureFileExists(filePath); err != nil {
+		lcs.Logger.ErrorWithCtx(ctx,
+			"Failed to ensure log file",
+			"runId", runId,
+			"filePath", filePath)
+		return
+	}
+
+	// get logs from pod, and keep the stream open (follow)
+	podLogOptions := &v1.PodLogOptions{
+		Follow: true,
+	}
+	restClientRequest := lcs.kubeClientSet.CoreV1().Pods(lcs.namespace).GetLogs(podName, podLogOptions)
+
+	// initialize stream and error for the while loop
+	var (
+		streamErr error
+		stream    io.ReadCloser
+	)
+
+	// stream logs - retry if failed
+	for {
+		stream, streamErr = restClientRequest.Stream(ctx)
+		if streamErr == nil {
+			break
+		}
+		lcs.Logger.WarnWithCtx(ctx, "Failed to get pod log read/closer", "runId", runId)
+		time.Sleep(1 * time.Second)
+	}
+	defer stream.Close() // nolint: errcheck
+
+	for {
+		buf := make([]byte, 1024)
```

Comment:
why allocating buffer everytime? this is a very expose memory / cpu operation.
you must be more gentle here. use bufferpool. we want to have constant memory consumption. buffer pool need to be configurable

Reply:
I'm thinking of using https://pkg.go.dev/sync#Pool,
but what do you want to be configurable? the buffer size?